rfcs/taskforce/0001-buildkiteci.md

```diff
@@ -0,0 +1,215 @@
+
+### Description of pipelines
+
+Pipeline are configured with Dhall rather than jinja-templated yaml. Dynamic pipelines support this as a first-class citizen, we do not require running the equivalent "`[test.py](http://test.py)`" before pushing. I have already implemented this on a [side project](https://github.com/bkase/gameboy/blob/master/.buildkite/pipeline.yaml).
+
+**General Pipeline Writing Things**
+
+* Use the [BuildKite monorepo plugin](https://github.com/chronotc/monorepo-diff-buildkite-plugin) for granular test execution.
+* Explicitly opt-in to the dependency graph. Don't support the "wait" instruction that blocks all parallel work.
+
+Assuming this proposal is accepted, more details will follow on how to structure the Dhall descriptions, but here is a rough cut of what it could look like:
+
+```haskell
+-- This is a definition of the package-deb step for Coda
+-- Optional decorations stack on top of the required bits
+-- This step will rerun twice before giving up
+-- It will upload artifacts under /tmp/artifacts/** to google cloud
+-- It depends on coda-build, genesis-ledger, and libp2p-helper steps
+-- Then the 4 required fields are given: a nice label for the UI, the
+--   command to run, a machine-readable key for further dependency graph
+--   manipulation, and a list of globs for granular test execution.
+-- Note: By making dirtyWhen a _required_ field, we set ourselves up for
+--       a world in which CI doesn't run on docs changes
+let packageDebStep =
+   decorate3
+	   (withFlaky 2)
+	   (withArtifacts "/tmp/artifacts/**")
+		 (withDeps [ "coda-build", "genesis-ledger", "libp2p-helper" ])
+	   Command::{
+	     label = ":box: Package deb",
+	     command = "make deb"
+	     key = "package-deb",
+	     dirtyWhen = [ "src/**" ]
+	   }
+```
+
+### Workflow of adding/changing/debugging pipelines
+
+Run buildkite agents locally or on manually spun up compute instances to test rules by targeting specific agents with tags and using the CLI [https://github.com/buildkite/cli](https://github.com/buildkite/cli)
+
+Liberal use of artifact uploading/downloading to keep build re-playable from different steps
+
+## This design solves the above issues
+
+P1 → self-hosted on GKE
+
+P2.1.1 → Parallel by default; opt-in to dependencies
+
+P2.1.2 → Rely on artifacts in gcloud which should be very quick to upload and download from within Gcloud infrastructure
+
+P2.2 → We can choose whatever machines we want
+
+P2.3 → We use dynamic pipelines to expand work based on files touched
+
+P3 → Manual adhoc template expansion is replaced with *automatic* expansion using Dhall
+
+P4 → BuildKite agents can be installed and tagged locally (or on your own instances) to iterate on builds on your own hardware before putting it into production
+
+P5 → Flakiness is explicit in the Dhall configuration
+
+## Mapping of resource utilization to tasks
```

Comment:
Which machines do unit tests run on? Depending on how we distribute the various unit tests across buildkite tasks, we may be able to get away with a light machine (since unit tests do not parallelize within a single library, only across multiple libraries containing unit tests).

Reply:
> Depending on how we distribute the various unit tests across buildkite tasks, we may be able to get away with a light machine (since unit tests do not parallelize within a single library, only across multiple libraries containing unit tests).

Distributing across unit tests will be a bit of extra effort because we'll need to schedule based on dirty transitive deps of a library rather than just on the library itself. I was thinking this can be future work (or at the least can be a project we decide to prioritize after we have the rest of the system in place)

Reply:
Hmm, yeah that is trickier and probably out of scope for the taskforce. FWIW, we could look into setting up https://github.com/Octachron/codept for our repo. It's not perfect yet, so we would need to play with it some to make it work for our repo, but it's the best ocaml dependency analyzer out there right now. If we had a graph of our deps, we should be able to have an easy script which returns the "dirty" libs for a commit.